### DISCUSSION TITLE
complementBed error w/ genome file

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1201] [Q&A] [answered]
## TL;DR `complementBed` doesn't like the format of my genome file and I'm not sure why

I'm using [complementBed](https://bedtools.readthedocs.io/en/latest/content/tools/complement.html) to generate a non-coding sequence track for the Roslin genome in [this Jupyter notebook](https://github.com/RobertsLab/project-gigas-oa-meth/blob/master/code/08-Generating-Genome-Feature-Tracks.ipynb). Once I have a non-coding sequence track, I can create an intron track.

When I try running the `complementBed` code (pulled from [my *C. virginica* notebook](https://github.com/epigeneticstoocean/paper-gonad-meth/blob/master/code/07-Generating-Genome-Feature-Tracks.ipynb)), I get the following error regarding [my genome file](https://github.com/RobertsLab/project-gigas-oa-meth/blob/master/genome-feature-files/cgigas_uk_roslin_v1-sequence-lengths.txt):

![Screen Shot 2021-05-05 at 10 37 53 AM](https://user-images.githubusercontent.com/22335838/117184772-f3f75a00-ad8d-11eb-9e9b-c23878a55548.png)

I have the two required fields: chromosome and chromosome length, and (I believe) it's a tab-delimited file! I used the following code in my Jupyter notebook to create the file:

![Screen Shot 2021-05-05 at 10 40 11 AM](https://user-images.githubusercontent.com/22335838/117185043-459fe480-ad8e-11eb-916f-978c52781d8f.png)

![Screen Shot 2021-05-05 at 10 40 24 AM](https://user-images.githubusercontent.com/22335838/117185082-4d5f8900-ad8e-11eb-8a61-26fa822a7288.png)

![Screen Shot 2021-05-05 at 10 40 40 AM](https://user-images.githubusercontent.com/22335838/117185115-57818780-ad8e-11eb-84ba-f43b92b11e4a.png)

**Comment by yaaminiv** [comment]
Realized `bedtools` v.2.17.0 was in my path, but I used v.2.26.0 in my previous code! I was able to switch to that version and run my code

**Reply by kubu4** [reply]
Whoa! Tricky! Nice troubleshooting!

How'd you catch the version differences?

**Reply by yaaminiv** [reply]
I looked at the directory I was calling with the `{bedtoolsDirectory}` variable in my previous notebook, and saw that it was different from the one in my path! So then I checked the versions.

**Comment by yaaminiv** [comment, marked as answer]
Update, I now have the same error with `flankBed`:

I want to create 1000 bp flanking regions for my gene track using the following code (pulled from [this notebook](https://github.com/hputnam/Meth_Compare/blob/master/code/03.01-Generating-Genome-Feature-Tracks.ipynb)):

![Screen Shot 2021-05-05 at 11 21 06 AM](https://user-images.githubusercontent.com/22335838/117190205-fd83c080-ad93-11eb-9ca1-6842670739be.png)

I'm using `flankBed` v.2.26.0 and encountering this error, but I'm not encountering that error for `complementBed` anymore! So now I am confused.

**Reply by kubu4** [reply]
I'd try to remake the genome BED file. I see that you've used the `paste` command and have added a delimiter of `"\t"`. However, the default delimiter used by `paste` _is_ a tab. Possible that instructing it to explicitly use a tab is screwing things up?

Also, code used in one of your previous notebooks to generate the genome bed file didn't utilize the extra `paste` step - it was done in a single `awk` command. Can you create your new genome bed file in the same fashion as that older example?

**Reply by yaaminiv** [reply]
> Possible that instructing it to explicitly use a tab is screwing things up?

I remade the genome BED file without adding `-d"\t"` and I got the same error!

> Can you create your new genome bed file in the same fashion as that older example?

I couldn't figure out how to do it in a single `awk` command! When I used the `awk` command from the other notebook, I get the additional descriptive text from the genome file (ex. "Crassostrea gigas linkage group 1, cgigas_uk_roslin_v1, whole genome shotgun sequence". I tried using `tr` to convert the spaces to tabs:

![Screen Shot 2021-05-05 at 11 46 43 AM](https://user-images.githubusercontent.com/22335838/117193333-923bed80-ad97-11eb-854e-2749b5553d0c.png)

![Screen Shot 2021-05-05 at 11 47 04 AM](https://user-images.githubusercontent.com/22335838/117193387-9e27af80-ad97-11eb-98d3-155f5ab50445.png)

Unfortunately, the sequence length is not in the same column throughout the genome file (column 12 in the first chunk, column 11 in the second chunk). Extracting the information separately and using `paste` was the only work around I could think of.

**Reply by yaaminiv** [reply]
Currently going down the rabbit hole of trying to cut the text string "Crassostrea.......shotgun sequence". I changed the `awk` command so it pulls out characters 2-14 of the sequence header. This works well for the end of the file, but not the beginning:

![Screen Shot 2021-05-05 at 12 08 11 PM](https://user-images.githubusercontent.com/22335838/117195779-90bff480-ad9a-11eb-97db-63cc10d14a89.png)

Is there a way to remove a variable number of characters from a string? Or specify that you want to remove strings that match an identifier (like somehow "grepping" out the string I don't want)

**Reply by kubu4** [reply]
I'd generate the file from the FastA index file instead of all that crazy `awk` stuff.

`cut -f1,2 GCF_902806645.1_cgigas_uk_roslin_v1_genomic.fna.fai`

**Reply by kubu4** [reply]
If you don't already have the FastA index files, you can generate with samtools:

`samtools faidx GCF_902806645.1_cgigas_uk_roslin_v1_genomic.fna`

**Reply by yaaminiv** [reply]
I ended up adding `sed 's/Cr//g'` after the `awk` code (take characters 2-14 of the header) to remove the remaining "Cr" in certain entries!

![Screen Shot 2021-05-05 at 12 28 10 PM](https://user-images.githubusercontent.com/22335838/117198029-5b68d600-ad9d-11eb-900e-8b3948fa1b08.png)

Still throws out the same error.

![Screen Shot 2021-05-05 at 12 28 29 PM](https://user-images.githubusercontent.com/22335838/117198064-67549800-ad9d-11eb-8a5b-2c4f452fca68.png)

I'll try using the index file and report back.

**Reply by yaaminiv** [reply]
OKAY NEW PROBLEM. I think `flankBed` is angry because the first line in my file is blank...

![Screen Shot 2021-05-05 at 12 33 35 PM](https://user-images.githubusercontent.com/22335838/117198640-1e511380-ad9e-11eb-8558-c2ffcab7b44d.png)

**Reply by kubu4** [reply]
I'd agree. Is that from the FastA index?

**Reply by yaaminiv** [reply]
Yup, that was the issue! I had a weird empty first line in my file.

![Screen Shot 2021-05-05 at 12 37 13 PM](https://user-images.githubusercontent.com/22335838/117199079-9e777900-ad9e-11eb-99ee-a5c3bf1abac8.png)

**Reply by yaaminiv** [reply]
> Is that from the FastA index?

No! I didn't end up using the FastA index
